tests/unit/s2n_mutual_auth_test.c

```diff
@@ -0,0 +1,349 @@
+
+        if (!cur_cipher->available) {
+            /* Skip Ciphers that aren't supported with the linked libcrypto */
+            continue;
+        }
+
+        server_cipher_preferences.suites = &cur_cipher;
+        config->cipher_preferences = &server_cipher_preferences;
+
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        EXPECT_SUCCESS(s2n_connection_set_client_auth_type(server_conn, S2N_CERT_AUTH_REQUIRED));
+        EXPECT_SUCCESS(s2n_connection_set_client_auth_type(client_conn, S2N_CERT_AUTH_REQUIRED));
+
+        /* Set up our I/O callbacks. Use stuffers for the "I/O context" */
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&client_to_server, 0));
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&server_to_client, 0));
+
+        /* Set Up Callbacks*/
+        EXPECT_SUCCESS(s2n_connection_set_recv_cb(client_conn, &buffer_read));
+        EXPECT_SUCCESS(s2n_connection_set_send_cb(client_conn, &buffer_write));
+        EXPECT_SUCCESS(s2n_connection_set_recv_cb(server_conn, &buffer_read));
+        EXPECT_SUCCESS(s2n_connection_set_send_cb(server_conn, &buffer_write));
+
+        /* Set up Callback Contexts to use stuffers */
+        EXPECT_SUCCESS(s2n_connection_set_recv_ctx(client_conn, &server_to_client));
+        EXPECT_SUCCESS(s2n_connection_set_send_ctx(client_conn, &client_to_server));
+        EXPECT_SUCCESS(s2n_connection_set_recv_ctx(server_conn, &client_to_server));
+        EXPECT_SUCCESS(s2n_connection_set_send_ctx(server_conn, &server_to_client));
+
+        int tries = 0;
+        do {
+            int ret;
+            ret = s2n_negotiate(client_conn, &client_blocked);
+            EXPECT_TRUE(ret == 0 || (client_blocked && errno == EAGAIN));
+            ret = s2n_negotiate(server_conn, &server_blocked);
+            EXPECT_TRUE(ret == 0 || (server_blocked && errno == EAGAIN));
+            tries += 1;
+
+            if (tries >= MAX_TRIES) {
+               FAIL();
+            }
+        } while (client_blocked || server_blocked);
+
+        /* Verify that both connections are waiting for Application Data */
+        EXPECT_TRUE(APPLICATION_DATA == s2n_conn_get_current_message_type(client_conn));
+        EXPECT_TRUE(APPLICATION_DATA == s2n_conn_get_current_message_type(server_conn));
+        EXPECT_TRUE(server_conn->handshake.handshake_type & CLIENT_AUTH);
+        EXPECT_TRUE(client_conn->handshake.handshake_type & CLIENT_AUTH);
+
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+    }
+
+
+    /*
+     * Test Mutual Auth using **s2n_config_set_client_auth_type**
+     */
+
+    EXPECT_SUCCESS(s2n_config_set_client_auth_type(config, S2N_CERT_AUTH_REQUIRED));
+
```

Comment:
Is there a test that checks what happens if auth type is set with both connection and config? Does one override the other?